sdk/identity/Azure.Identity/src/TokenExchangeManagedIdentitySource.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.Identity
+{
+    internal class TokenExchangeManagedIdentitySource : ManagedIdentitySource
+    {
+        private TokenFileCache _tokenFileCache;
+        private ClientAssertionCredential _clientAssertionCredential;
+
+        private TokenExchangeManagedIdentitySource(CredentialPipeline pipeline, string tenantId, string clientId, string tokenFilePath)
+            : base(pipeline)
+        {
+            _tokenFileCache = new TokenFileCache(tokenFilePath);
+            _clientAssertionCredential = new ClientAssertionCredential(tenantId, clientId, _tokenFileCache.GetTokenFileContents, new ClientAssertionCredentialOptions { Pipeline = pipeline });
+        }
+
+        public static ManagedIdentitySource TryCreate(CredentialPipeline pipeline, string clientId)
+        {
+            string tokenFilePath = EnvironmentVariables.AzureFederatedTokenFile;
```

Comment:
should we throw if the path is not found or we can't access the file there?

Reply:
This will throw when we call `File.ReadAllText`. I don't think we should try to check that the path exists or we have access up front as this could change between when we do the check and when we read the file, so I don't think the check would provide much value.

Reply:
I think the only value add is that we'd potentially fail earlier here, but agreed it's not a big value add.